keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,579 @@
+
+ ```
+	builder := &utils.NetworkPolicySpecBuilder{}
+	builder = builder.SetName("allow-x-via-pod-and-ns-selector").SetPodSelector(map[string]string{"pod": "a"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod":"b"}, map[string]string{"ns":"y"}, nil, nil)
+
+	reachability := utils.NewReachability(allPods, true)
+	reachability.ExpectAllIngress(Pod("x/a"), false)
+	reachability.Expect(Pod("y/b"), Pod("x/a"), true)
+	reachability.Expect(Pod("x/a"), Pod("x/a"), true)
+  ```
+ This represents a significant reduction in code complexity, with the equivalent tests using the existing `network_policy.go` implementation being 3 to 4 times as long, mostly due to boilerplate around verification and go structures.
+
+*Further improvements to the testing API*
+
+- Make the function calls in network policy builder *even* more DSL-like, for example, 
+```
+Pod(...).InNamespace(...).CanAccess(...)
+```
+- Infer `Egress,Ingress` rules rather than force them to be specified, based on builder inputs.  They're redundant to begin with (i.e. calico doesn't even require them)
+- Add `From` and `To` semantics to the struct API calls in reachability.
+ 
+###  Note on Acceptance and Backwards compatibility
+
+Thus far there are two obvious ways to ensure backwards compatibility. 
+- Each old test will be converted to a truth table, first, as part of this work, so that the parity between old and new tests is clear and obvious 
+- Alternatively, these can be a next generation of policy tests which live in parallel to existing network policy tests for a release cycle, while they are vetted
+
+## Next steps
+
+As of now, network policy tests are not run regularly against any CNI.  Although we should not endorse one CNI over another, we should regularly validate
+that the NetworkPolicy tests *can* pass on *some* provider.  As part of this proposal, we propose commiting an annotation to the existing `network_policy.go` code which states, in clear and simple terms, what environment the `network_policy.go` test suite was run in, the last time which it was committed and passed.  It's also acceptable to commit this as a Markdown file in the documentation.
+ 
+There may be other, better ways of doing this.  Running an upstream validation job of these tests as a weekly PROW job, for example, would be a good way to make sure that these tests don't regress in the future.  This comes at the cost of coupling a job to an external CNI provider, so its not being explicitly suggested.
+
+## Other Improvement Ideas
+
```

Comment:
Another suggestion: test if there is an ordering dependency. In other words, there are distinct actions:

1. create server pod
2. create policy
3. create client pod
4. set labels (on namespace / pod, if not set on creation) 
5. evaluate policy

It would be good to test that these operations are invariant on order; in other words, we should be able to re-label pods or create pods before / after policy creation, and still pass validation.

Reply:
yeah, good point - two ways to do this:

1) Actually permute the ordering : logically complex and lots of permutations to test.

2) stochastically  by doing more of a scale test
- create 30 pods, and 30 policies in go routines
- continue probing until all pods are running, for all policies to match probe statuses

since some folks wanted (2) anyways, do you feel like that is a good way to fulfill the requirement?

Reply:
cc @abhiraut

Reply:
Would be interesting to try permutations, my concern there would be that the tests themselves would take forever to run ... so yeah, leaning towards @jayunit100 's suggestion, stochastically could still flush out bugs without blowing up the runtime too much!

Reply:
i dont intend to have the permutations on all test cases.. but have specific test case or two for this purpose